src/Stratis.Features.FederatedPeg/TargetChain/InputConsolidator.cs

```diff
@@ -0,0 +1,250 @@
+                    return ConsolidationSignatureResult.Failed();
+                }
+
+                this.logger.LogDebug("Successfully signed transaction.");
+
+                // NOTE: We don't need to reserve the transaction. The wallet will be at a standstill whilst this is happening.
+
+                // If it is FullySigned, broadcast.
+                if (this.walletManager.ValidateConsolidatingTransaction(this.PartialTransaction, true))
+                {
+                    this.logger.LogDebug("Consolidation transaction is fully signed. Broadcasting {0}", this.PartialTransaction.GetHash());
+                    this.broadcasterManager.BroadcastTransactionAsync(this.PartialTransaction);
+                    this.fullySigned = true;
+                }
+
+                return ConsolidationSignatureResult.Succeeded(this.PartialTransaction);
+            }
+        }
+
+        /// <summary>
+        /// Build a consolidating transaction.
+        /// </summary>
+        public Transaction BuildConsolidatingTransaction()
+        {
+            try
+            {
+                // TODO: Confifm that we can remove the ordering below.
+
+                List<UnspentOutputReference> unspentOutputs = this.walletManager.GetSpendableTransactionsInWallet(WithdrawalTransactionBuilder.MinConfirmations).ToList();
+
+                if (unspentOutputs.Count < WithdrawalTransactionBuilder.MaxInputs)
+                    throw new Exception("We shouldn't be consolidating transactions if we have less than 50 UTXOs to spend.");
+
+                IEnumerable<UnspentOutputReference> orderedUnspentOutputs = DeterministicCoinOrdering.GetOrderedUnspentOutputs(unspentOutputs);
+                IEnumerable<UnspentOutputReference> selectedInputs = orderedUnspentOutputs.Take(WithdrawalTransactionBuilder.MaxInputs);
+                string walletPassword = this.walletManager.Secret.WalletPassword;
+                bool sign = (walletPassword ?? "") != "";
+
+                var multiSigContext = new TransactionBuildContext(new List<Recipient>())
+                {
+                    MinConfirmations = WithdrawalTransactionBuilder.MinConfirmations,
+                    Shuffle = false,
+                    IgnoreVerify = true,
+                    WalletPassword = walletPassword,
+                    Sign = sign,
+                    TransactionFee = ConsolidationFee,
+                    SelectedInputs = selectedInputs.Select(u => u.ToOutPoint()).ToList(),
+                    AllowOtherInputs = false,
+                    IsConsolidatingTransaction = true
+                };
+
+                Transaction transaction = this.transactionHandler.BuildTransaction(multiSigContext);
+
+                this.logger.LogDebug("Consolidating transaction = {0}", transaction.ToString(this.network, RawFormat.BlockExplorer));
+
+                return transaction;
+            }
+            catch (Exception e)
```

Comment:
Ah I see now that we catch the exception here.. still I would perhaps return something else from the method or just log the false if statement and return null up top.

Reply:
This code or the idea of it is mostly copied straight from WithdrawalTransactionBuilder.

The reason we are catching exceptions is the same as for there - there may be a NotEnoughFunds exception or similar so we do need the try catch in case the wallet's state changes before we start doing this. 

I can definitely change it to return null above instead of throwing an exception though, that makes sense. Does that sound better?